pkg/hub/rbacfinalizerdeletion/controller.go

```diff
@@ -0,0 +1,240 @@
+	clusterv1 "github.com/open-cluster-management/api/cluster/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	rbacv1informers "k8s.io/client-go/informers/rbac/v1"
+	rbacv1client "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const (
+	manifestWorkFinalizer = "cluster.open-cluster-management.io/manifest-work-cleanup"
+)
+
+type finalizeController struct {
+	roleLister         rbacv1listers.RoleLister
+	roleBindingLister  rbacv1listers.RoleBindingLister
+	rbacClient         rbacv1client.RbacV1Interface
+	clusterLister      clusterv1listers.ManagedClusterLister
+	namespaceLister    corelisters.NamespaceLister
+	manifestWorkLister worklister.ManifestWorkLister
+	eventRecorder      events.Recorder
+}
+
+// NewFinalizeController ensures all manifestworks are deleted before role/rolebinding for work
+// agent are deleted in a terminating cluster namespace.
+func NewFinalizeController(
+	roleInformer rbacv1informers.RoleInformer,
+	roleBindingInformer rbacv1informers.RoleBindingInformer,
+	namespaceLister corelisters.NamespaceLister,
+	clusterLister clusterv1listers.ManagedClusterLister,
+	manifestWorkLister worklister.ManifestWorkLister,
+	rbacClient rbacv1client.RbacV1Interface,
+	eventRecorder events.Recorder,
+) factory.Controller {
+
+	controller := &finalizeController{
+		roleLister:         roleInformer.Lister(),
+		roleBindingLister:  roleBindingInformer.Lister(),
+		namespaceLister:    namespaceLister,
+		clusterLister:      clusterLister,
+		manifestWorkLister: manifestWorkLister,
+		rbacClient:         rbacClient,
+		eventRecorder:      eventRecorder,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, roleInformer.Informer(), roleBindingInformer.Informer()).
+		WithSync(controller.sync).ToController("FinalizeController", eventRecorder)
+}
+
+func (m *finalizeController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	key := controllerContext.QueueKey()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore role/rolebinding whose key is not in format: namespace/name
+		return nil
+	}
+
+	cluster, err := m.clusterLister.Get(namespace)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	ns, err := m.namespaceLister.Get(namespace)
+	if err != nil {
+		return err
+	}
+
+	// cluster is deleted or being deleted
+	role, rolebinding, err := m.getRoleAndRoleBinding(namespace, name)
```

Comment:
Also note, if you had a `CriticalResource` you wouldn't be writing this code at all and we would be inverting the checks.